system/jlib/jmetrics.cpp

```diff
@@ -0,0 +1,219 @@
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jmetrics.hpp"
+#include "jmutex.hpp"
+#include "jlog.hpp"
+
+using namespace hpccMetrics;
+
+static Singleton<MetricsReporter> metricsReporter;
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+
+MODULE_EXIT()
+{
+    delete metricsReporter.queryExisting();
+}
+
+
+MetricsReporter &hpccMetrics::queryMetricsReporter()
+{
+    return *metricsReporter.query([] { return new MetricsReporter; });
+}
+
+
+void CounterMetric::collect(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<uint32_t>>(name, description, ValueType::METRICS_UNSIGNED, count.load())));
+}
+
+
+void GaugeMetric::collect(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<float>>(name, description, ValueType::METRICS_INTEGER, gaugeValue.load())));
+}
+
+
+MetricsReporter::~MetricsReporter()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        delete sinkIt.second.pSink;
+    }
+}
+
+
+void MetricsReporter::init(IPropertyTree *pMetricsTree)
+{
+    Owned<IPropertyTree> pSinkTree = pMetricsTree->getPropTree("sinks");
+    auto sinkElementsIt = pSinkTree->getElements("sink");
+    initializeSinks(sinkElementsIt);
+}
+
+
+void MetricsReporter::addMetric(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(pMetric->queryName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->queryName(), pMetric});
+    }
+    // todo - need to decide how to handle duplicate named metrics
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(metricName);
+    if (it != metrics.end())
+    {
+        metrics.erase(it);
+    }
+}
+
+
+void MetricsReporter::startCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->startCollection(this);
+    }
+}
+
+
+void MetricsReporter::stopCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->stopCollection();
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::queryMetricsForReport(const std::string &sinkName)
+{
+    std::vector<std::shared_ptr<IMetric>> reportMetrics;
+
+    auto it = sinks.find(sinkName);
+    if (it != sinks.end())
```

Comment:
else throw an internal error.
Also - with this implementation the name can't be optional.